File: implementations/go/examples/internal/application/command/cancel_payment_handler.go
package command

import (
	"context"

	"github.com/example/account-service/internal/domain/payment"
)

type CancelPaymentCommand struct {
	PaymentID   string
	Reason      string
	RequesterID string
}

// CancelPaymentHandler processes a payment cancellation. The PaymentCancelled
// Domain Event raised by Payment.Cancel() is translated into a
// payment.cancelled.v1 Integration Event, which the Account BC subscribes to
// in order to run a compensating credit (reversing the amount already debited).
type CancelPaymentHandler struct {
	repo payment.Repository
}

func NewCancelPaymentHandler(repo payment.Repository) *CancelPaymentHandler {
	return &CancelPaymentHandler{repo: repo}
}

// Handle saves and returns immediately — publishing to/consuming from SQS via
// the Outbox is solely the responsibility of the independently, periodically
// running outbox.Poller/outbox.Consumer (no synchronous draining,
// domain-events.md).
func (h *CancelPaymentHandler) Handle(ctx context.Context, cmd CancelPaymentCommand) (*payment.Payment, error) {
	p, err := payment.FindOne(ctx, h.repo, cmd.PaymentID, cmd.RequesterID)
	if err != nil {
		return nil, err
	}

	if err := p.Cancel(cmd.Reason); err != nil {
		return nil, err
	}

	if err := h.repo.SavePayment(ctx, p); err != nil {
		return nil, err
	}
	return p, nil
}

File: implementations/fastapi/examples/src/payment/application/command/cancel_payment_handler.py
from dataclasses import dataclass

from ...domain.errors import PaymentNotFoundError
from ...domain.payment import Payment
from ...domain.payment_repository import PaymentRepository


@dataclass
class CancelPaymentCommand:
    requester_id: str
    payment_id: str
    reason: str


class CancelPaymentHandler:
    def __init__(self, repo: PaymentRepository) -> None:
        self._repo = repo

    async def execute(self, cmd: CancelPaymentCommand) -> Payment:
        payments, _ = await self._repo.find_payments(
            page=0, take=1, payment_id=cmd.payment_id, owner_id=cmd.requester_id
        )
        payment = payments[0] if payments else None
        if payment is None:
            raise PaymentNotFoundError(cmd.payment_id)

        payment.cancel(cmd.reason)
        await self._repo.save_payment(payment)
        # PaymentCancelled → the Account BC subscribes to payment.cancelled.v1 and executes
        # the compensating credit — handled asynchronously by OutboxPoller/OutboxConsumer (domain-events.md).
        return payment
